File: lib/c/cos_sim.c

#define BILLION 1000000000.0


double inner_prod(double *a, double *z) {
    double value = 0.0;
    while (a != z) {
        value += ( (*a) * (*a) );
        ++a;
    }
    return value;
}

double get_euclidean_magnitude(int size, double *a, double *b) {

    double a_len, b_len;
    int i;

    a_len = inner_prod(a, a+size);
    b_len = inner_prod(b, b+size);

    /*
    for (i=0; i<size; i++) {
        a_len +=(a[i]*a[i]);
    }
    for (i=0; i<size; i++) {
        b_len += (b[i]*b[i]);
    }
    */

    return sqrt(a_len * b_len);
}


double get_dot_product(int size, double *a, double *b) {

    double dot_prod;
    int i;
    
    dot_prod = 0;
    i = 0;
    
    for (i=0; i<size; i++) {
        dot_prod += (a[i] * b[i]);
    }
    return dot_prod;
}


//double get_cosine_similarity(int size, double *a, double *b, double *eucl_t, double *dot_t) {
double get_cosine_similarity(int size, double *a, double *b) {

    double eucl_magn, dot_prod, cos_sim;
    //struct timespec start, end;

    //clock_gettime(CLOCK_REALTIME, &start);
    eucl_magn = get_euclidean_magnitude(size, a, b);
    //clock_gettime(CLOCK_REALTIME, &end);
    //*eucl_t += (end.tv_sec - start.tv_sec) + ((end.tv_nsec - start.tv_nsec)/BILLION);

    //clock_gettime(CLOCK_REALTIME, &start);
    dot_prod = get_dot_product(size, a, b);
    //clock_gettime(CLOCK_REALTIME, &end);
    //*dot_t += (end.tv_sec - start.tv_sec) + ((end.tv_nsec - start.tv_nsec)/BILLION);

    if (eucl_magn == 0.0) {
        cos_sim = 0.0;
    } else {
        cos_sim = dot_prod / eucl_magn;
    }

    return cos_sim;
}

File: pkg/py/cos_sim.py

from math import sqrt

def get_cosine_similarity(A, B):
    """
    Cosine similarity calculates the angle distance between two n dimensional
    vectors. The length of the vectors and therefore the eucledian distance
    between them does not play a role, i.e. if the vectors represent documents,
    a longer documents might be still similar to a shorter one.

    The similarity value is 1 if the vectors are identical (i.e. have the same
    angle), 0 if the are not similar (vectors are perpendicular) and -1 if the
    vectors have opposite directions.
    """

    dot_prod = get_dot_prod(A,B)
    eucl_magn = get_eucl_magn(A,B)
    return dot_prod / eucl_magn if eucl_magn else None

def get_dot_prod(A,B):
    """
    Dot product represents how much to vectors have in common, includes the
    angle and length of these vectors. The value is positive if they have
    some similarity, 0 if there is no similarity at all and negative if they
    are opposite of each other.
    """
    dot_prod = 0
    for a, b in zip(A, B):
        dot_prod += a * b
    return dot_prod


def get_eucl_magn(A,B):
    """
    Eucledian magnitude is the product of the lengths of the two vectors.
    We use this value to normalize the dot product of the vectors and
    neutralize the effect of vector lengths on the final similarity score.
    """
    A_len = 0
    B_len = 0
    for a, b in zip(A,B):
        A_len += a * a
        B_len += b * b
    return sqrt(A_len * B_len)
